src/openssl/async_to_sync_wrapper.rs

```diff
@@ -0,0 +1,87 @@
+use std::fmt::{Debug, Formatter};
+use std::io;
+use std::io::Read;
+use std::io::Write;
+use std::pin::Pin;
+use std::task::Context;
+use std::task::Poll;
+
+use futures_lite::io::{AsyncRead, AsyncWrite};
+
+pub(super) struct AsyncToSyncWrapper<S> {
+    inner: S,
+    context: usize,
+}
+
+impl<S> AsyncToSyncWrapper<S>
+where
+    S: Unpin,
+{
+    /// Wrap sync object in this wrapper.
+    pub fn new(inner: S) -> Self {
+        Self { inner, context: 0 }
+    }
+
+    /// Store async context inside this object
+    pub fn set_context(&mut self, cx: &mut Context<'_>) {
+        assert_eq!(self.context, 0);
+        self.context = cx as *mut _ as usize;
+    }
+
+    /// Clear async context
+    pub fn unset_context(&mut self) {
+        assert_ne!(self.context, 0);
+        self.context = 0;
+    }
+
+    fn with_context<F, R>(&mut self, f: F) -> R
+    where
+        F: FnOnce(&mut Context<'_>, Pin<&mut S>) -> R,
+    {
+        unsafe {
```

Comment:
any work around unsafe?

Reply:
No alternative for this one. All we have is a pointer to work with.